### PR TITLE
Hook: LIST.updateAll

### PR DESCRIPTION
**Pull request by andreiburdusa**
---
Part of #2253 

~~Needs answer for: https://runtimeverification.slack.com/archives/CC360GUTG/p1605864619247900?thread_ts=1605699058.240400&cid=CC360GUTG~~

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: docs/hooks.md [hunk 1]
```diff
@@ -704,6 +704,18 @@ The number of elements in the given list.
         [hook{}("LIST.size")]
 ~~~
 
+## LIST.updateAll
+
+`LIST.updateAll(L1, index, L2)` creates a new list using a list `L2` of size `N`
+where the `N` elements of `L1` starting at `index` are replaced with the
+contents of `L2`. The result is `\bottom` if `index` is negative or part of `L2`
+would go of the end of `L1`.
```
Comment (ana-pantilie) [inline docs/hooks.md:712]: Could you reword this somehow? I'm having trouble understanding what `part of L2
would go of the end of L1` means.

Comment (andreiburdusa) [reply docs/hooks.md:712]: I used [this](https://runtimeverification.slack.com/archives/CC360GUTG/p1605892398253700?thread_ts=1605699058.240400&cid=CC360GUTG) explanation. I'll try to find a clearer one.

File: docs/hooks.md [hunk 2]
```diff
@@ -704,6 +704,21 @@ The number of elements in the given list.
...
 
+## LIST.updateAll
+
+`LIST.updateAll(L1, index, L2)` creates a new list using a list `L2` of size `N`
+where the `N` elements of `L1` starting at `index` are replaced with the
+contents of `L2`.
+The result is `\bottom` in these cases:
+
+- `index` is negative
+- `L2` is non-empty and `index + N` exeeds the lenth of `L1`
```
Comment (ana-pantilie) [inline docs/hooks.md:715]:
```suggestion
- `L2` is non-empty and `index + N` exeeds the length of `L1`
```

File: kore/test/Test/Kore/Builtin/List.hs [hunk 3]
```diff
@@ -417,6 +418,36 @@ test_size =
...
+            let new = asInternal . fmap mkInt $ Seq.fromList [1, 2, 3, 4]
+            result <-
+                evaluate
+                $ updateAllList original (mkInt 0) new
+            assertEqual' "" Pattern.bottom result
+    ]
+  where
+    original = asInternal . fmap mkInt $ Seq.fromList [1, 2, 3]
+
+
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Builtin/List.hs:450]:
```suggestion
```

File: kore/src/Kore/Builtin/List.hs [hunk 4]
```diff
@@ -324,6 +327,27 @@ evalSize resultSort [_list] = do
...
+            | _ix < 0 = return (Pattern.bottomOf resultSort)
+            | len2 == 0 = returnList resultSort _list1
+            | _ix + len2 > len1 = return (Pattern.bottomOf resultSort)
+            | otherwise =
+                returnList
+                    resultSort
+                    ( Seq.take _ix _list1
+                        <> _list2
+                        <> Seq.drop (_ix + length _list2) _list1
+                    )
```
Comment (ana-pantilie) [inline kore/src/Kore/Builtin/List.hs:347]: Would using intermediate results declared in a let binding make the a little code clearer here?

**Review by ana-pantilie — APPROVED** [review]

**Pull request merged**